src/TSTransformer/nodes/statements/transformForOfStatement.ts

```diff
@@ -422,27 +422,26 @@ const buildGeneratorLoop: LoopBuilder = makeForLoopBuilder((state, initializer,
 });
 
 function getLoopBuilder(state: TransformState, node: ts.Node, type: ts.Type): LoopBuilder {
-	if (isDefinitelyType(type, isArrayType(state))) {
+	if (isDefinitelyType(state, type, node, isArrayType(state))) {
 		return buildArrayLoop;
-	} else if (isDefinitelyType(type, isSetType(state))) {
+	} else if (isDefinitelyType(state, type, node, isSetType(state))) {
 		return buildSetLoop;
-	} else if (isDefinitelyType(type, isMapType(state))) {
+	} else if (isDefinitelyType(state, type, node, isMapType(state))) {
 		return buildMapLoop;
-	} else if (isDefinitelyType(type, t => isStringType(t))) {
+	} else if (isDefinitelyType(state, type, node, isStringType)) {
 		return buildStringLoop;
-	} else if (isDefinitelyType(type, isIterableFunctionLuaTupleType(state))) {
+	} else if (isDefinitelyType(state, type, node, isIterableFunctionLuaTupleType(state))) {
 		return buildIterableFunctionLuaTupleLoop(type);
-	} else if (isDefinitelyType(type, isIterableFunctionType(state))) {
+	} else if (isDefinitelyType(state, type, node, isIterableFunctionType(state))) {
 		return buildIterableFunctionLoop;
-	} else if (isDefinitelyType(type, isGeneratorType(state))) {
+	} else if (isDefinitelyType(state, type, node, isGeneratorType(state))) {
 		return buildGeneratorLoop;
-	} else if (isDefinitelyType(type, isIterableType(state))) {
+	} else if (isDefinitelyType(state, type, node, isIterableType(state))) {
 		DiagnosticService.addDiagnostic(errors.noIterableIteration(node));
 		return () => luau.list.make();
-	} else if (type.isUnion()) {
-		DiagnosticService.addDiagnostic(errors.noMacroUnion(node));
-		return () => luau.list.make();
 	} else {
+		// TODO: Check for the type's `[Symbol.iterator]` property and use that
```

Comment:
Please file a design issue for this instead of leaving a comment in the compiler.

Reply:
I felt like this was not design, but fundamental breakage relating to types. TS will allow it, but the typechecks don't cover it so it assert fails.

Reply:
But removed half of the comment. Left in the part about known breakage.